### ISSUE TITLE
Debugging: step over behavior

### ISSUE THREAD

**Issue by krangerich** [dbalsom/martypc#80]
All the debugging capabilities are awesome!

But I have some additional suggestions regarding the "step over" button in the "CPU Control" window:
- Currently, during "rep" instructions (rep movsw, ...), step over and step into behavior is the same. It would be very convenient, when "step over" would skip the loop and jump to the next instruction
- Currently, if you "step over" a long call / int and a hardware int occurs, the debugger stops after the iret of the hardware interrupt (=somewhere randomly inside the long call / int function). I don't know if this is indented, but it would be more convenient, if it would reliably end up at the next instruction after the call / int you want to "step over"

**Comment by dbalsom** [comment]
> All the debugging capabilities are awesome!

Hi! Glad you find them useful.

> * Currently, during "rep" instructions (rep movsw, ...), step over and step into behavior is the same. It would be very convenient, when "step over" would skip the loop and jump to the next instruction

Agreed it would be nice to be able to skip reps and loops and such. Step over currently only has an effect on calls but if the current instruction is not a call I could just have it skip to the next decoded address.

> * Currently, if you "step over" a long call / int and a hardware int occurs, the debugger stops after the iret of the hardware interrupt (=somewhere randomly inside the long call / int function). I don't know if this is indented, but it would be more convenient, if it would reliably end up at the next instruction after the call / int you want to "step over"

So step over is a bit of a hack at the moment. The main issue is that there's no guarantee that you can step over anything. Internally I am saving the return address on a call, but we may not end up back at that address for various reasons, interrupts being one.   So there is a timeout - if we don't return after a certain number of instructions I just break after the timeout elapses.  This ends up plopping you in some random place.   Without the timeout, you might just step over something and see the emulator just keep running again forever after.  I wasn't sure which method is better in the end. 

Do you know how other emulators/debuggers handle this?

**Comment by krangerich** [comment]
I just checked debugging Java in IntelliJ IDEA as a random example. When you hit "step over" on a method and the method does not return (for example, there could just be an infinite loop), the program just keeps running and does not return to the debugger (unless another breakpoint is hit).

I think, this is simple and reasonable behavior for a debugger. If the function call / method call / .. never returns, you just have to start over again and use "step into" instead (or change breakpoints accordingly).
Basically you could act, as if there was a breakpoint on the following instruction.

**Comment by krangerich** [comment]
Hey, thanks, thats good to hear!
I'm not able to try it yet, because I'm currently busy on the hardware side (assembling a Xi8088, ...)
